tests/models/test_taskinstance.py

```diff
@@ -992,3 +994,103 @@ def success_handler(self, context):
+
+    def _test_previous_dates(self, schedule_interval: Union[str, datetime.timedelta, None], catchup: bool,
+                             object_: str) -> None:
+        dag_id = 'test_previous_dates'
+        dag = models.DAG(dag_id=dag_id, schedule_interval=schedule_interval, catchup=catchup)
+        task = DummyOperator(task_id='task', dag=dag, start_date=DEFAULT_DATE)
+
+        d1 = '2019-01-01T00:00:00+00:00'
+        d2 = '2019-01-02T00:00:00+00:00'
+        d3 = '2019-01-03T00:00:00+00:00'
+
+        with create_session() as session:  # type: Session
+
+            # session.query(DagRun).all().filter(DagRun.dag_id == dag_id).all()
+            # session.query(TI).filter(TI.dag_id == dag_id).delete()
+            # session.commit()
+            dag.create_dagrun(
+                run_id='scheduled__{}'.format(d1),
+                state=State.SUCCESS,
+                execution_date=pendulum.parse(d1),
+                start_date=pendulum.utcnow(),
+                session=session
+            )
+
+            dag.create_dagrun(
+                run_id='scheduled__{}'.format(d2),
+                state=State.FAILED,
+                execution_date=pendulum.parse(d2),
+                start_date=pendulum.utcnow(),
+                session=session
+            )
+
+            dag.create_dagrun(
+                run_id='scheduled__{}'.format(d3),
+                state=State.SUCCESS,
+                execution_date=pendulum.parse(d3),
+                start_date=pendulum.utcnow(),
+                session=session
+            )
+
+            t1 = TI(task=task, execution_date=pendulum.parse(d1))
+            t1.set_state(state=State.SUCCESS, session=session)
+
+            t2 = TI(task=task, execution_date=pendulum.parse(d2))
+            t2.set_state(state=State.SUCCESS, session=session)
+
+            t3 = TI(task=task, execution_date=pendulum.parse(d3))
+            t3.set_state(state=State.SUCCESS, session=session)
+
+            if object_ == 'previous_ti':
+                self.assertIsNone(t1.previous_ti)
+                assert isinstance(t3.previous_ti, TI)  # mypy
+                self.assertEqual(t3.previous_ti.execution_date, pendulum.parse(d2))
+
+            elif object_ == 'previous_ti_success':
+                self.assertIsNone(t1.previous_ti_success)
+                assert isinstance(t3.previous_ti_success, TI)  # mypy
+                self.assertEqual(t3.previous_ti_success.execution_date, pendulum.parse(d1))
+
+            elif object_ == 'previous_execution_date_success':
+                self.assertIsNone(t1.previous_execution_date_success)
+                self.assertEqual(t3.previous_execution_date_success, pendulum.parse(d1))
+
+            elif object_ == 'previous_start_date_success':
+                self.assertIsNone(t1.previous_start_date_success)
+                self.assertEqual(t3.previous_start_date_success, t1.start_date)
+
+            self.tearDown()
+
+    def test_previous_ti(self) -> None:
```

Comment:
These test would be better using `@parameterized` - check out test/jobs/test_scheduler_job.py for an example.